Classes/Emogrifier/HtmlProcessor/AbstractHtmlProcessor.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+namespace Pelago\Emogrifier\HtmlProcessor;
+
+/**
+ * Base class for HTML processor that e.g., can remove, add or modify nodes or attributes.
+ *
+ * @author Oliver Klee <github@oliverklee.de>
+ */
+abstract class AbstractHtmlProcessor
+{
+    /**
+     * @var string
+     */
+    const DEFAULT_DOCUMENT_TYPE = '<!DOCTYPE html>';
+
+    /**
+     * @var string
+     */
+    const CONTENT_TYPE_META_TAG = '<meta http-equiv="Content-Type" content="text/html; charset=utf-8">';
+
+    /**
+     * @var \DOMDocument
+     */
+    private $xmlDocument = null;
+
+    /**
+     * @param string $unprocessedHtml raw HTML, will get heavily normalized
+     *
+     * @throws \InvalidArgumentException if $unprocessedHtml is anything other than a non-empty string
+     */
+    public function __construct($unprocessedHtml)
+    {
+        if (!is_string($unprocessedHtml)) {
+            throw new \InvalidArgumentException('The provided HTML must be a string.', 1515459744321);
+        }
+        if ($unprocessedHtml === '') {
+            throw new \InvalidArgumentException('The provided HTML must not be empty.', 1515763647038);
+        }
+
+        $this->xmlDocument = $this->createXmlDocument($unprocessedHtml);
+    }
+
+    /**
+     * Renders the normalized and processed HTML.
+     *
+     * @return string
+     */
+    public function render()
```

Comment:
As far as I can conclude, child classes of this abstract class will need to override public method `render()` and call `parent::render()` in each implementation.

I would introduce a protected method `process()`, abstract in this base class, which would be responsible for holding the variable logic of each subclass (I think this is called the "template design pattern", but it's not that important):

(And maybe even declare the `render()` method `final`.)

```php
abstract protected function process();

...

final public function render()
{
    $this->process();
    return $this->xmlDocument->saveHTML();
}
```

What do you think?

Reply:
You are right, it makes sense. For some reason, I was under the impression that the `convertCssToAttributes()`, `removeClassAttributes()` etc. methods will only set flags for what there is to do and then the `render()` method will do the actual job.